Enable a linter rule to prevent usage of "import" in README.md javascript code blocks

WeiJun428 opened the pull request:
This is a PR that is intended to solve issue #15746.

Summary of What I did:
- ran `rush add -p eslint-plugin-markdown --dev` in `common\tools\eslint-plugin-azure-sdk`
- modified `common\tools\eslint-plugin-azure-sdk\src\configs\azure-sdk-base.ts`:
  - included the `eslint-plugin-markdown` plugin, 
  - added `override` configuration that splits typescript and javascript linting. 
  - used `no-restricted-import` to inhibit ES6 import usage.
- added `README.md` as target and removed `--ext` option of lint script in `sdk\textanalytics\ai-text-analytics\package.json` and fixed the existing error.

jeremymeng commented:
I tried `npx eslint README.md --debug` and found that we are using `javascript` tag for the code block.  
>   eslint:linter A code block was found: '17.javascript' +0ms

so we probably need to support both *.js and *.javascript in the overrides config.

There's also another issue. Looks that we need to change eslint to allow parsing .javascript extension.

> The file does not match your project config: README.md/20_20.javascript.
> The extension for the file (.javascript) is non-standard. It should be added to your existing "parserOptions.extraFileExtensions"

WeiJun428 commented:
Thank you, @jeremymeng!

In my recent commits, I:
- enabled `javascript` extension in `parserOptions/extraFileExtension`
- supported `javascript` in overrides config by using `"**/*.md/*.{js,javascript}"`
- factored out typescript specific `parserOptions/project`, `extends`, and `rules` in `override` to avoid applying type-specific rules to JavaScript code, according to ["StackOverflow: parserOptions.project" has been set for @typescript-eslint/parser](https://stackoverflow.com/questions/58510287/parseroptions-project-has-been-set-for-typescript-eslint-parser)
- added a custom error message to `no-restricted-import`: "Please use require instead of import."
- changed `--ext .ts` to `--ext .ts,.md,.javascript,.js` in lint scripts in `sdk\textanalytics\ai-text-analytics\package.json` to enable markdown file linting

The only thing left to enable the rule in every sdk is adding README.md and `--ext .ts,.md,.javascript,.js` to `lint:` and `lint:fix` in `package.json` of every sdk root repo.

WeiJun428 commented:
@jeremymeng, I have fixed the errors and resolved the conflicts in my recent commits.

jeremymeng commented on sdk/textanalytics/ai-text-analytics/package.json:
```diff
@@ -70,8 +70,8 @@
-    "lint:fix": "eslint package.json api-extractor.json src test --ext .ts --fix --fix-type [problem,suggestion]",
-    "lint": "eslint package.json api-extractor.json src test --ext .ts",
+    "lint:fix": "eslint README.md package.json api-extractor.json src test --ext .ts,.md,.javascript,.js --fix --fix-type [problem,suggestion]",
+    "lint": "eslint README.md package.json api-extractor.json src test --ext .ts,.md,.javascript,.js",
```
curious whether `.md,.javascript,.js` is required in `--ext`.  Previously we could lint package.json without specifying `.json` extension

WeiJun428 replied:
I tried to experiment with different `--ext` arguments. It turns out that the eslint in only linting the snippet when either
- `--ext` is not defined at all. It is because ESLint v7 [automatically lints file extensions specified in overrides[].files patterns in config files](https://github.com/eslint/rfcs/blob/0253e3a95511c65d622eaa387eb73f824249b467/designs/2019-additional-lint-targets/README.md). It looks like eslint encourages not using `--ext` options when linting a wide array of file types as it is redundant with the `override` property.
- `--ext .md,.javascript` is included.

This also brings up another question where we might only want to keep `.javascript` instead of keeping both `.javascript` and `.js`, but only with the promise that all javascript sinppets in `README.md` are named `javascript`. I tried to look over multiple `README.md` and this turns out to be true, but I am still unsure whether it holds true for the whole repo.

jeremymeng replied:
we discussed and decided to remove `--ext`

jeremymeng reviewed (commented):
The split in azure-sdk-base makes sense to me since we don't want to to use ts rules on javascript.

@deyaaeldeen @witemple-msft could you please review too?

jeremymeng commented:
@WeiJun428 please update PR description as some comments are outdated now.  Thanks!

deyaaeldeen commented on common/config/rush/pnpm-lock.yaml:
```diff
@@ -4963,6 +5012,10 @@ packages:
       is-extglob: 2.1.1
     dev: false
 
+  /is-hexadecimal/1.0.4:
```
why is this needed? same question for most of the deps being added here.

WeiJun428 replied:
I believe these are added by the new plugin `eslint-plugin-markdown`

deyaaeldeen reviewed (commented):
It looks good other than the issues highlighted in existing comments. Thanks for your contribution!

WeiJun428 commented on common/tools/eslint-plugin-azure-sdk/src/configs/azure-sdk-base.ts:
```diff
@@ -4,124 +4,151 @@
+        "plugin:markdown/recommended"
+      ],
+      rules: {
+        "no-restricted-imports": ["error", {
```
README.md is not added as a target to lint in most `package.json` except in text-analytics, hence putting it as error will not raise any errors unless we try to fix the errors in subsequent PRs.

WeiJun428 commented:
/check-enforcer override

jeremymeng commented:
Looking good now! @WeiJun428 please update PR title and description.

WeiJun428 commented:
@jeremymeng updated!

witemple-msft commented:
The code and changes to the linter configuration look good. Could you give this issue a descriptive title? Something like "[eslint-plugin] Support linting README.md and checking for `require` instead of `import`"?

jeremymeng commented:
> The code and changes to the linter configuration look good. Could you give this issue a descriptive title? Something like "[eslint-plugin] Support linting README.md and checking for `require` instead of `import`"?

Oops sorry @witemple-msft I didn't see your comments when I tweaked the title then merged this.
